components/kyma-environment-broker/internal/orchestration/strategy_parallel.go

```diff
@@ -11,42 +10,61 @@ import (
 )
 
 type ParallelOrchestrationStrategy struct {
-	operationExecutor process.Executor
-	log               logrus.FieldLogger
+	executor process.Executor
+	log      logrus.FieldLogger
 }
 
-// TODO(upgrade): Finish implementation and write tests; unused for now
-func NewParallelOrchestrationStrategy(operationExecutor process.Executor, log logrus.FieldLogger) Strategy {
+func NewParallelOrchestrationStrategy(executor process.Executor, log logrus.FieldLogger) Strategy {
 	return &ParallelOrchestrationStrategy{
-		operationExecutor: operationExecutor,
-		log:               log,
+		executor: executor,
+		log:      log,
 	}
 }
 
 func (p *ParallelOrchestrationStrategy) Execute(operations []internal.RuntimeOperation, strategySpec internal.StrategySpec) (time.Duration, error) {
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
+	if len(operations) == 0 {
+		return 0, nil
+	}
 
-	q := process.NewQueue(p.operationExecutor, p.log)
+	stopCh := make(chan struct{})
 
-	q.Run(ctx.Done(), strategySpec.Parallel.Workers)
+	q := process.NewQueue(p.executor, p.log)
+	q.Run(stopCh, strategySpec.Parallel.Workers)
 
-	isMaintenanceWindowMode := strategySpec.Schedule == internal.MaintenanceWindow
+	isMaintenanceWindowSchedule := strategySpec.Schedule == internal.MaintenanceWindow
 
-	// Sort operations according to TimeDelta(Now, internal.RuntimeOperation.MaintenanceWindowBegin)
-	if isMaintenanceWindowMode {
+	if isMaintenanceWindowSchedule {
+		for i, op := range operations {
+			operations[i].MaintenanceWindowBegin = p.resolveWindowTime(op.MaintenanceWindowBegin, op.MaintenanceWindowEnd)
+		}
 		sort.Slice(operations, func(i, j int) bool {
 			return operations[i].MaintenanceWindowBegin.Before(operations[j].MaintenanceWindowBegin)
 		})
 	}
 
 	for _, op := range operations {
-		if isMaintenanceWindowMode {
-			time.Sleep(time.Until(op.MaintenanceWindowEnd))
+		if isMaintenanceWindowSchedule {
+			until := time.Until(op.MaintenanceWindowBegin)
+			p.log.Infof("Waiting %v to start upgrade operation %s", until, op.Operation.ID)
+			time.Sleep(until)
```

Comment:
This logic doesn't guarantee that all operations will be executed within the requested time window. 
Think of an example where typical operation takes 30 minutes. Executing 10 operations with 2 workers would take ~150 minutes. If time window is is the same 1 hour for each runtime, tbis logic executes beyond the time frame.

Reply:
I have an idea how to implement scheduling properly by relying on the underlying `DelayedQueue`'s `AddAfter()` method. PLease ping me to discuss.